muxdb/cache.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2021 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package muxdb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/qianbin/directcache"
+	"github.com/vechain/thor/v2/trie"
+)
+
+// cache is the cache layer for trie.
+type cache struct {
+	queriedNodes   *directcache.Cache // caches recently queried node blobs.
+	committedNodes *directcache.Cache // caches newly committed node blobs.
+	roots          struct {           // caches root nodes.
+		m        map[string]trie.Node
+		lock     sync.Mutex
+		maxMajor uint32
+		ttl      uint32
+	}
+
+	nodeStats   cacheStats
+	rootStats   cacheStats
+	lastLogTime int64
+}
+
+// newCache creates a cache object with the given cache size.
+func newCache(sizeMB int, rootTTL uint32) *cache {
+	sizeBytes := sizeMB * 1024 * 1024
+	cache := &cache{
+		queriedNodes:   directcache.New(sizeBytes / 4),
+		committedNodes: directcache.New(sizeBytes - sizeBytes/4),
+		lastLogTime:    time.Now().UnixNano(),
+	}
+	cache.roots.m = make(map[string]trie.Node)
+	cache.roots.ttl = rootTTL
+	return cache
+}
+
+func (c *cache) log() {
+	now := time.Now().UnixNano()
+	last := atomic.SwapInt64(&c.lastLogTime, now)
+
+	if now-last > int64(time.Second*20) {
+		log1, ok1 := c.nodeStats.ShouldLog("node cache stats")
+		log2, ok2 := c.rootStats.ShouldLog("root cache stats")
+
+		if ok1 || ok2 {
+			log1()
+			log2()
+		}
+	} else {
+		atomic.CompareAndSwapInt64(&c.lastLogTime, now, last)
+	}
+}
+
+// AddNodeBlob adds encoded node blob into the cache.
+func (c *cache) AddNodeBlob(keyBuf *[]byte, name string, path []byte, ver trie.Version, blob []byte, isCommitting bool) {
```

Comment:
Looks like most of these methods are exported but not used outside the package. Maybe I'm wrong ? Probably doesn't matter too much anyhow

Reply:
Itself can be treated a standalone "class" to other ones, even it's inside the same package. IMHO, it's a good practice to define public/private functions even it's internal only.